src/main/java/org/testng/internal/ClassImpl.java

```diff
@@ -100,13 +102,15 @@ private Object getDefaultInstance(boolean create, String errMsgPrefix) {
       if (m_instance != null) {
         m_defaultInstance = m_instance;
       } else {
-        Object instance = getInstanceFromGuice();
-
-        if (instance != null) {
-          m_defaultInstance = instance;
+        IObjectProvider provider = ObjectProviderFactory.newInstance(m_testContext, m_class);
+        if (provider.supportsDependencyInjection()) {
```

Comment:
I think this logic should be managed by the provider itself.

Reply:
@juherr - My idea was to have providers just be able to plugin their implementations and let TestNG decide on how it would want to use them. The heavy lifting in terms of the logic would still reside within TestNG. That is why I chose to take this approach. I am not sure if it would make sense for us to give away the complete control of object creation to the interface implementors

Reply:
I disagree because the design is not clear enough: the separation between the logic and its usage is not clear enough.

If you want, we can make it step by step but the API must not be public: it will allow us to redesign it without backward compatibility in mind.

Reply:
@juherr - I didn't quite understand your observation. Can you please elaborate? I merely following the same approach of other implementations within TestNG wherein the end user plugins their implementation, but TestNG decides how to use it and when to use it. 

The provider just provides ways of creating objects for the two means. TestNG decides when to invoke what, based on the decisions that it internally makes. Why should that logic reside within the provider ? Can you help me understand that part ?

Reply:
Ping @juherr

Reply:
In fact, I don't catch why it should have 2 ways for object creations.
If the need for `supportsDependencyInjection` is real then I think it should not be exposed to the SPI.
Maybe an intermediate layer is missing.

At the same time, the current change is a good step too.

What do you think if you move everything under the internal package and we will be able to make it public later?

Reply:
@juherr 

I decided to give this a fresh look. Revamped the implementation and have force pushed the changes. Can you please start a fresh look at this PR ? 

Note: I have moved the implementation to an internal package as suggested by you.